Collation number and some fixes

hwwhww opened the pull request:
## Description

Sorry about that this PR contains many different sides during I was writing simulation.

1. Added `get_shard_list function` in validator manager contract

2. Added a `ShardChain.sync()` function for simulating fast sync.

3. Added `number` field to `CollationHeader` for some reasons:
    1. It’s easier to delete old junk collations in database with `CollationHeader.number` since it’s easier to locate the old collation by their number.
    2. It’s easier to **peek** the score of an orphan collation with `CollationHeader.number`:
it would help the shard validators to determine if they fell behind when the given collation’s parent is **not** in database, and then they can ask for the parent collations. Otherwise, the shard validators have to call validator manager contract to check the score **after** they received the block which includes the given collation.

    And it would taking more space for storing this one new field and some cost of switching shard branch would increase by additional data management: if we want to use a function like `ShardChain.get_collation_by_number(collation_number)`, it may takes more computational time upon receiving a new block and reorganization the collation head since we have to reorganize database values like: `self.db.put(b'coll:%d' % collation.header.number, collation.header.hash)`

4. Since `self.main_chain` was been assigned as a member variable in `ShardChain` in previous commit, I reuse it for add_collation arguments. Though IMO it still makes the relationship of `ShardChain` and `MainChain` unclear.

5. Fixed some pep8 and flake8 checks

6. A bugfix for `test_receipt_consuming_tx_utils.mk_testing_receipt_consuming_tx`
    * Use argument `startgas` instead of `t.GASPRICE`
https://github.com/ethereum/sharding/commit/2af298c68487133f79b29eea276f144372149093#diff-bffb8f12854c82f71a1017109ee4b1cbR32
    * Not sure of the logic of this test:
https://github.com/ethereum/sharding/commit/2af298c68487133f79b29eea276f144372149093#diff-bffb8f12854c82f71a1017109ee4b1cbR116 and will discuss with @mhchia .

7. Updated `get_validators_max_index` function in validator manager contract to prevent inconsistent shard_list. Note that the deposit would be activated since next shuffling cycle.

8. Separate some basic utils functions from `validator_manager_utils.py` -> `contract_utils.py`

9. Rename `SHUFFLING_CYCLE`  to `SHUFFLING_CYCLE_LENGTH`

10. Set SHUFFLING_CYCLE_LENGTH from 2500 to 25 for now (easier for testing before EIP96 is finished)

11. Updated tester function call: added `is_constant` flag so it can be used for calling constant function and it won’t change the head state

## TODO (next PR)
1. Implement `ShardChain.get_collation_by_number(collation_number)` function
2. Delete old junk collations

## Cute Animal Picture
![polar_bear](https://media.giphy.com/media/fSHgyXYXy8Sd2/giphy.gif)

mhchia commented on sharding/validator_manager_utils.py:
```diff
@@ -165,35 +165,48 @@ def call_contract_inconstantly(state, ct, contract_addr, func, args, value=0, st
+    # return abi.decode_abi(['bool[100]'], shard_list)[0]
+
+
+def call_tx_add_header(state, sender_privkey, value, header, gasprice=GASPRICE, nonce=None, startgas=300000):
```
Why do we add `nonce` as a parameter here? It seems not used yet.

hwwhww replied:
Ahh, it was used for handling validator's txqueue, but now I use another way to implement.
So you are right, I don't need to assign `nonce` now. I'll fix it. 👍

mhchia commented on sharding/main_chain.py:
```diff
@@ -1,9 +1,21 @@
 from ethereum.pow.chain import Chain
+from ethereum.utils import (
+    sha3, hash32,
+    big_endian_to_int, encode_hex)
```
should this be reordered?

hwwhww replied:
okay! 👍

mhchia commented on sharding/tests/test_validator_manager_utils.py:
```diff
@@ -3,18 +3,21 @@
+    get_shard_list,
+    mk_validation_code,
+    sign,
+    create_contract_tx,
```
should this be reordered as well?

hwwhww replied:
okay!

hwwhww commented on sharding/tests/test_validator_manager.py:
```diff
@@ -5,14 +5,15 @@
+    mk_validation_code,
+    sign,
+    WITHDRAW_HASH,
+    DEPOSIT_SIZE,
```
Do you mean put `WITHDRAW_HASH ` and `DEPOSIT_SIZE` to top? No problem. :)

mhchia replied:
yeah thank you

hwwhww commented:
@mhchia 
As working on simulation, I updated the control flow of receipt-consuming-tx.
Please review commit ada2bf8c66c3b48c911872ee4f8074fbf5d32f30 :)

1. Updated and rename `is_valid_receipt_consuming_tx` to  `validate_receipt_consuming_tx`: this function raises exceptions for clear log information
2. Calling `validate_receipt_consuming_tx` in `send_msg_transfer_value`. The usage of `validate_receipt_consuming_tx` is similar to the usage of `validate_transaction` in `apply_transaction`.
3. Collators don’t include the receipt-consuming-txs that can’t pass `validate_receipt_consuming_tx`. Since only collators would call `add_transactions` while they’re creating a new collation, I modify the `add_transactions`.

Thanks!

mhchia commented on sharding/receipt_consuming_tx_utils.py:
```diff
@@ -77,6 +93,8 @@ def send_msg_add_used_receipt(state, shard_id, receipt_id):
 
 
 def send_msg_transfer_value(mainchain_state, shard_state, shard_id, tx):
+    assert validate_receipt_consuming_tx(mainchain_state, shard_state, shard_id, tx)
```
IMO, the `assert` should only be used in the scenario that an event should always be `True` or `False` in our program.
For example, `validate_receipt_consuming_tx` might throw some exception like `InvalidTransaction` when we receive an ill-formed transaction from the network.

hwwhww replied:
Ahhh, sorry, it's a careless mistake.
It is supposed to be only:
```
def send_msg_transfer_value(mainchain_state, shard_state, shard_id, tx):
    validate_receipt_consuming_tx(mainchain_state, shard_state, shard_id, tx)
```
And the exceptions from `add_shard_transaction` from `apply_collation` from `add_collation` will be caught.

mhchia replied:
Get it. Thank you:)
